### PR TITLE
feat: Support FreeParameter expressions in AutoQASM programs

### PR DESCRIPTION
**Pull request by rmshaffer**
*Issue #, if available:*
[AutoQASM: support for FreeParameterExpressions](https://github.com/orgs/amazon-braket/projects/2/views/1?pane=issue&itemId=43127738)

*Description of changes:*
Support `FreeParameterExpression` objects as inputs to AutoQASM programs and subroutines, in addition to `FreeParameter` objects.

*Testing done:*
Unit tests added. tox passes.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/experimental/autoqasm/program/program.py [hunk 1]
```diff
@@ -322,22 +323,21 @@ def register_args(self, args: list[Any]) -> None:
...
-                # TODO laurecap: Support for expressions
-                raise NotImplementedError(
-                    "Expressions of FreeParameters will be supported shortly!"
-                )
+                for expression_arg in preorder_traversal(arg._expression.args):
+                    if isinstance(expression_arg, Symbol):
+                        self.register_parameter(expression_arg.name)
 
-    def register_parameter(self, parameter: FreeParameter) -> None:
+    def register_parameter(self, parameter_name: str) -> None:
```
Comment (math411) [inline src/braket/experimental/autoqasm/program/program.py:332]: Could this accept a union of str, FreeParam?

Then L339, 340 can check for repr(parameter_name) and add that value in.

File: src/braket/experimental/autoqasm/program/program.py [hunk 2]
```diff
@@ -322,22 +323,21 @@ def register_args(self, args: list[Any]) -> None:
...
         for arg in args:
             if isinstance(arg, FreeParameter):
-                self.register_parameter(arg)
+                self.register_parameter(arg.name)
             elif isinstance(arg, FreeParameterExpression):
-                # TODO laurecap: Support for expressions
-                raise NotImplementedError(
-                    "Expressions of FreeParameters will be supported shortly!"
-                )
+                for expression_arg in preorder_traversal(arg._expression.args):
```
Comment (math411) [inline src/braket/experimental/autoqasm/program/program.py:328]:
```
if isinstance(arg, FreeParameterExpression):
    for expression_arg in preorder_traversal(arg._expression.expr_free_symbols):
```

Comment (math411) [reply src/braket/experimental/autoqasm/program/program.py:328]: Then expression_arg can just be cast to a `str`

Comment (rmshaffer) [reply src/braket/experimental/autoqasm/program/program.py]: It gives a warning that `.expr_free_symbols` is deprecated, so I have used `.free_symbols` instead. Much cleaner this way, thank you. I do still have to check that it is a `Symbol`, as `.free_symbols` can contain other items (e.g. constants) which are not `Symbol`s.

Comment (math411) [reply src/braket/experimental/autoqasm/program/program.py]: Can you use this to simplify the if, elif statement? `FreeParameter` should have a `._expression` to use, well.

File: test/unit_tests/braket/experimental/autoqasm/test_parameters.py [hunk 3]
```diff
@@ -775,3 +785,139 @@ def parametric(val: float):
...
+
+
+def test_sim_expressions():
+    @aq.main
+    def parametric():
+        rx(0, 2 * FreeParameter("phi"))
+        measure(0)
+
+    measurements = _test_parametric_on_local_sim(parametric(), {"phi": np.pi / 2})
+    assert 0 not in measurements["__bit_0__"]
```
Comment (laurencap) [inline test/unit_tests/braket/experimental/autoqasm/test_parameters.py:812]: woohoo

File: src/braket/experimental/autoqasm/program/program.py [hunk 4]
```diff
@@ -321,40 +323,61 @@ def register_args(self, args: list[Any]) -> None:
...
         """
         for arg in args:
-            if isinstance(arg, FreeParameter):
-                self.register_parameter(arg)
-            elif isinstance(arg, FreeParameterExpression):
-                # TODO laurecap: Support for expressions
-                raise NotImplementedError(
-                    "Expressions of FreeParameters will be supported shortly!"
-                )
+            if isinstance(arg, FreeParameterExpression):
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/program/program.py:326]: nice!

File: src/braket/experimental/autoqasm/program/program.py [hunk 5]
```diff
@@ -321,40 +323,61 @@ def register_args(self, args: list[Any]) -> None:
...
+            expr (FreeParameterExpression): The expression in which to look for free symbols.
+
+        Returns:
+            Iterable[str]: The list of free symbol names in sorted order (sorted to ensure
+            that the order is deterministic).
+        """
+        return sorted([str(s) for s in expr._expression.free_symbols if isinstance(s, Symbol)])
 
-    def register_parameter(self, parameter: FreeParameter) -> None:
+    def register_parameter(self, parameter: str | FreeParameter) -> None:
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/program/program.py:344]: Should we make this accept only strings? Originally it had been that way.

File: src/braket/experimental/autoqasm/program/program.py [hunk 6]
```diff
@@ -321,40 +323,61 @@ def register_args(self, args: list[Any]) -> None:
...
+            not already been registered with the program.
 
         Returns:
-            FloatVar: The associated variable.
+            FloatVar: The variable representing the expression.
         """
-        if name not in self._free_parameters:
-            raise errors.ParameterNotFoundError(f"Free parameter '{name}' was not found.")
-        return self._free_parameters[name]
+        # Validate that all of the free symbols are registered as free parameters.
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/program/program.py:368]: Nice solution!

**Review by laurencap — APPROVED** [review]

**Review by yitchen-tim — APPROVED** [review]

File: src/braket/experimental/autoqasm/program/program.py (hunk 5 above)
Comment (rmshaffer) [reply src/braket/experimental/autoqasm/program/program.py:344]: Yes, in fact it is only called with strings now. Will update.

File: src/braket/experimental/autoqasm/program/program.py [hunk 7]
```diff
@@ -321,40 +323,60 @@ def register_args(self, args: list[Any]) -> None:
...
+        which are Symbol objects.
+
+        Args:
+            expr (FreeParameterExpression): The expression in which to look for free symbols.
+
+        Returns:
+            Iterable[str]: The list of free symbol names in sorted order (sorted to ensure
+            that the order is deterministic).
+        """
+        return sorted([str(s) for s in expr._expression.free_symbols if isinstance(s, Symbol)])
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply src/braket/experimental/autoqasm/program/program.py:342]: Yes, for example I noticed that integer literals are part of `free_symbols`. See: https://docs.sympy.org/latest/modules/core.html#sympy.core.basic.Basic.free_symbols

**Pull request merged**